flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/AdaptiveJoinProcessor.java

```diff
@@ -0,0 +1,170 @@
+                                    i,
+                                    ExecEdge.builder()
+                                            .source(newNode)
+                                            .target(node)
+                                            .shuffle(edge.getShuffle())
+                                            .exchangeMode(edge.getExchangeMode())
+                                            .build());
+                        }
+                    }
+                };
+
+        List<ExecNode<?>> newRootNodes =
+                execGraph.getRootNodes().stream()
+                        .map(
+                                node -> {
+                                    node = replaceAdaptiveJoinNode(node);
+                                    node.accept(visitor);
+                                    return node;
+                                })
+                        .collect(Collectors.toList());
+
+        return new ExecNodeGraph(execGraph.getFlinkVersion(), newRootNodes);
+    }
+
+    private ExecNode<?> replaceAdaptiveJoinNode(ExecNode<?> node) {
+        if (!(checkAllInputShuffleIsHash(node))
+                || isUpstreamNodeKeepInputAsIs(node.getInputEdges())) {
+            return node;
+        }
+        ExecNode<?> newNode = node;
+        if (node instanceof AdaptiveJoinExecNode
+                && ((AdaptiveJoinExecNode) node).canBeTransformedToAdaptiveJoin()) {
+            BatchExecAdaptiveJoin adaptiveJoin = ((AdaptiveJoinExecNode) node).toAdaptiveJoinNode();
+            replaceInputEdge(adaptiveJoin, node);
+            newNode = adaptiveJoin;
+        }
+
+        return newNode;
+    }
+
+    private boolean checkKeepInputAsIsExisted(List<InputProperty> inputProperties) {
+        return inputProperties.stream()
+                .anyMatch(
+                        inputProperty ->
+                                inputProperty.getRequiredDistribution().getType()
+                                        == KEEP_INPUT_AS_IS);
+    }
+
+    private boolean isUpstreamNodeKeepInputAsIs(List<ExecEdge> inputEdges) {
+        return inputEdges.stream()
+                .filter(execEdge -> execEdge.getSource() instanceof BatchExecExchange)
+                .map(execEdge -> (BatchExecExchange) execEdge.getSource())
+                .anyMatch(exchange -> checkKeepInputAsIsExisted(exchange.getInputProperties()));
+    }
+
+    private boolean isAdaptiveJoinEnabled(ProcessorContext context) {
+        TableConfig tableConfig = context.getPlanner().getTableConfig();
+        boolean isAdaptiveJoinEnabled =
+                tableConfig.get(
+                                        OptimizerConfigOptions
+                                                .TABLE_OPTIMIZER_ADAPTIVE_BROADCAST_JOIN_STRATEGY)
+                                != OptimizerConfigOptions.AdaptiveBroadcastJoinStrategy.NONE
+                        && !TableConfigUtils.isOperatorDisabled(
```

Comment:
This should not affect SkewedJoin optimization.

Reply:
Yes, SkewedJoin optimization will add a 'or' condition, so it wont be affected.